Upgrade tokio

phlmn opened the pull request:
This is the PR for #44 based on the work of @izderadicka.

izderadicka committed:
Applied PR #43 for native-tls 0.2

izderadicka committed:
All tests running on linux

izderadicka committed:
examples  working except hyper server

izderadicka committed:
hyper examples working, but still some problem with server

izderadicka committed:
hyper examples working, but still some problem with server

izderadicka committed:
Merge branch 'new-tokio' of https://github.com/izderadicka/tokio-tls into new-tokio

phlmn committed:
Do not crash when connection fails

phlmn committed:
Remove tokio-proto feature

izderadicka commented:
@phlmn 
Thanks for help.  Server is not halting after error ( I guess there is no better way, as for_each will stop on first error as I now learned). However the server solution is still suboptimal.
First -  I had to use panic! in the error handling, because I cannot find way how to map error due to this problem:
```
error[E0271]: type mismatch resolving `<[closure@examples/hyper-server.rs:66:49: 66:83] as std::ops::FnOnce<(hyper::Error,)>>::Output == hyper::common::never::Never`
  --> examples/hyper-server.rs:66:26
   |
66 |                     conn.and_then(|c| c.map_err(|e| eprintln!("Hyper error {}", e)))
   |                          ^^^^^^^^ expected (), found enum `hyper::common::never::Never`
   |
   = note: expected type `()`
              found type `hyper::common::never::Never`
   = note: required because of the requirements on the impl of `futures::Future` for `futures::MapErr<hyper::server::conn::Connection<tokio_tls::TlsStream<tokio::net::TcpStream>, hyper::service::service::ServiceFnOk<[closure@examples/hyper-server.rs:39:40: 39:87], hyper::Body>>, [closure@examples/hyper-server.rs:66:49: 66:83]>`

error[E0599]: no method named `map_err` found for type `futures::AndThen<hyper::server::conn::Connecting<tokio_tls::TlsStream<tokio::net::TcpStream>, futures::FutureResult<hyper::service::service::ServiceFnOk<[closure@examples/hyper-server.rs:39:40: 39:87], hyper::Body>, hyper::common::never::Never>>, futures::MapErr<hyper::server::conn::Connection<tokio_tls::TlsStream<tokio::net::TcpStream>, hyper::service::service::ServiceFnOk<[closure@examples/hyper-server.rs:39:40: 39:87], hyper::Body>>, [closure@examples/hyper-server.rs:66:49: 66:83]>, [closure@examples/hyper-server.rs:66:35: 66:84]>` in the current scope
```

Isn't there better way how to handle errors here?

Also the server is not performing well - I tried to run ` wrk -c100 -t10 -d20s https://localhost:12345` and throughput is very low (on linux with openssl):
```
Running 20s test @ https://localhost:12345
  10 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    73.63ms  231.94ms   1.99s    91.80%
    Req/Sec     1.31k     1.10k    3.53k    49.47%
  62486 requests in 20.09s, 5.18MB read
  Socket errors: connect 0, read 0, write 0, timeout 8
Requests/sec:   3110.63
Transfer/sec:    264.28KB
```

For comparison running on local nginx with TLS with same parameters I can get about 50k req per sec.

I guess it can be due to there errors in server (there are more like this):
```
thread 'tokio-runtime-worker-0' panicked at 'Hyper error error shutting down connection: Broken pipe (os error 32)', examples/hyper-server.rs:66:53
Error: error accepting connection: unexpected EOF
```

izderadicka commented:
Also one small change is needed remove `cargo test --feature tokio-proto` from travis and appveyor test scripts - this should hopefully make CI checks succeed

izderadicka committed:
Fix CI scripts

phlmn committed:
CI: Remove tests with tokio-proto feature

phlmn commented:
@izderadicka I've removed the CI tests with the tokio-proto feature. I will take a look at the error handling later. I also noticed the poor performance… How does it compare to the version with tokio-proto?

izderadicka commented:
Sorry I was idiot - I tested with debug build, with release build situation looks much better - performance slightly better then nginx, which is expected as test response is smaller:
```
wrk -c100 -t10 -d20s https://localhost:12345
Running 20s test @ https://localhost:12345
  10 threads and 100 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     9.87ms   40.43ms 513.53ms   94.96%
    Req/Sec     8.57k     3.42k   27.04k    86.86%
  1669585 requests in 20.09s, 138.52MB read
Requests/sec:  83086.14
Transfer/sec:      6.89MB
```

However the variance and max is still high -  I think it's due to threads panic, when wrk closes connection - see above my note about panic. Errors in server are:
```
tokio-runtime-worker-2' panicked at 'Hyper error error shutting down connection: Broken pipe (os error 32)', examples/hyper-server.rs:66:53
```

Also builds fine on linux but there is some issue OSX version - but I cannot do anything about it as I do not use Mac.

phlmn committed:
Update dependencies

phlmn committed:
Fix tests on mac

phlmn committed:
Try to fix mac test on travis

phlmn committed:
Fix mac test on travis

phlmn commented:
I fixed the tests failing on mac. There was an ugly problem when travis failed but it ran on my local mac.

The assertions in bad.rs and google.rs were checking if the error message contains `The trust policy was not trusted.` but on my local machine it was `The certificate was not trusted.` but on travis it still was `The trust policy was not trusted.`. I assume this is dependent on the native security-framework version installed on the system.

phlmn commented:
@alexcrichton Is there some way to run the tests on windows? The AppVeyor link seems broken.

alexcrichton commented:
I don't managed this repository any more, unfortunately

djc commented:
FWIW, I looked over the changes here and it looks good to me. The changes in `src` are very small. There's quite a bit of churn in the examples and the tests, but nothing that looks surprising.

carllerche commented on examples/hyper-client.rs:
```diff
@@ -64,39 +66,65 @@ struct HttpsConnector {
+
 }
+
+// impl Service for HttpsConnector {
```
If this is no longer needed, can it be deleted?

carllerche reviewed (changes requested):
Thanks for doing this and sorry for the delay.

I did a first pass review mostly focused on the lib code. Comments inline as well as some thoughts:

* impl `Future` for `MidHandshake`, `AsyncAccept`, (probably others) needs to be bound by `AsyncRead` and `AsyncWrite`. There seem to be TODOs in the code.

mati865 commented:
@phlmn It's been 2 weeks since review, will you have time to address the remarks?

izderadicka committed:
Merge branch 'new-tokio' of https://github.com/phlmn/tokio-tls into new-tokio

izderadicka committed:
Added changes required by PR review

Implemented TODOs to change binds to AsyncRead, AsyncWrite

changed version to 0.2.0

moved tokio to   dev-dependecies

added exact futures version to dependencies

removed unused code in examples/hyper-client

remove proto.rs - as it's not used anymore

updated readme - removed references to tokio-proto,  updated example.

izderadicka commented:
@phlmn -  I added changes required by Carl as PR to your branch see phlmn/tokio-tls#1 - Can you accept.

izderadicka committed:
updated readme

phlmn committed:
Merge pull request #1 from izderadicka/new-tokio

Changes required in PR review

phlmn commented:
Sorry, I've been very busy lately. I just merged @izderadicka's changes, is there anything left to do?

miquels commented:
Just as a data point, I built a crate that uses tokio-tls 0.2.0 with native-tls-0.2.1 and so far it works fine. Please publish :)  ( the crate is https://github.com/miquels/hyper-tls-hack )

seanmonstar reviewed (approved):
It looks like everything has been addressed, and CI is happy!
